### ISSUE TITLE
"pgsql" schema is not showing up in postgres dataset 

### ISSUE THREAD

**Issue by NavinAsokan** [apache/superset#30678]
### Bug description

The bug is that when I successfully created a new postgres connection, All the schemas are listed except the schema name "pgsql". This is a legacy system if we change the schema name then this will be a massive change from  the database backend. Superset version is 3.1.0

### Screenshots/recordings

<img width="328" alt="Screenshot 2024-10-22 at 2 31 02 PM" src="https://github.com/user-attachments/assets/6a85a8f5-8791-4b4c-b78d-a443fccbba5b">
<img width="446" alt="Screenshot 2024-10-22 at 2 33 11 PM" src="https://github.com/user-attachments/assets/d5ad7eb9-6112-4935-b375-2308366761a4">

### Superset version

master / latest-dev

### Python version

3.11

### Node version

I don't know

### Browser

Safari

### Additional context

 Superset version is 3.1.0

### Checklist

- [X] I have searched Superset docs and Slack and didn't find a solution to my problem.
- [X] I have searched the GitHub issue tracker and didn't find a similar bug report.
- [X] I have checked Superset's logs for errors and if I found a relevant Python stacktrace, I included it here as text in the "additional context" section.

**Comment by NavinAsokan** [comment]
Hi, Please share if we need to modify the manager.py under superset installation

**Comment by NavinAsokan** [comment]
please share the steps to assign this permission using superset user interface

**Comment by NavinAsokan** [comment]
We are unable to see the access to schema "pgsql" in the permission list. Remaining other schemas are visible. Pls suggest

**Comment by NavinAsokan** [comment]
For the below option, I am not getting the rights to assign schema for pgsql to user. I can see rights to assign other schemas that are visible which is the cause for raising this bug. The pgsql schema is setup the same way as other schemas. No change in setup. Pls suggest

Schema Permissions: Ensure that the "pgsql" schema has the correct permissions set in Superset. You can verify this by checking the role permissions in the Superset UI under Security > List Roles. Make sure the role associated with the user has the [database].[schema] permission for "pgsql".

**Comment by NavinAsokan** [comment]
Restarted superset . Still the issue persist.
Checked the logs. Logs didn't provide much information on this

**Comment by villebro** [comment]
Postgres is very widely used in Superset, so I would be very surprised if this is a bug in either the Postgres connector or Superset.

A few thoughts:
- Check that you have adequate permissions in Superset. Start with a user that has Admin role to ensure it's not a perm issue.
- Validate your connection string for the referenced catalog. Are you sure the catalog in Superset is the same as the one you're using in the other database IDE?

**Comment by NavinAsokan** [comment]
Thanks for the update. Yes. I am using Admin permission to login to superset and using the same connection string as that of the database IDE. The logs didn't provide much information as well while listing the dataset postgres schemas in superset. Is there any way we can provide more verbose info from the log?

**Comment by villebro** [comment]
Can you post a redacted copy of your connection string in Superset and one for the IDE? Also, does the schema dropdown show other schemas? And do you see an error toast after it stops loading?

**Comment by NavinAsokan** [comment]
From superset : postgresql+psycopg2://db_name:XXXXXXXXXX@hostname:port#/username
	From IDE: url: jdbc:postgresql://hostname:port#/db_name
<img width="1040" alt="Screenshot 2024-10-30 at 4 33 01 PM" src="https://github.com/user-attachments/assets/53febe2b-f35f-4cb0-b6de-14bdbfc64cf4">

**Comment by villebro** [comment]
I see you have `username` catalog in the Superset URL, but `db_name` in the IDE. If you're referencing different catalogs, you'll likely not find the schemas you see in `db_name` in the `username` catalog. IIUC, your Superset connection string should be

```
postgresql+psycopg2://username:XXXXXXXXXX@hostname:port#/db_name
```
(make sure the pwd is also updated to reflect `username` creds)

**Comment by NavinAsokan** [comment]
My bad on the representation for superset url. I wrongly mentioned that it is in the other format. We already have this format

postgresql+psycopg2://username:XXXXXXXXXX@hostname:port#/db_name. 

Still unable to get the pgsql schema name in the schema list. Kindly suggest

**Comment by NavinAsokan** [comment]
double check that username and db_name are the same in both Superset and IDE -- Checked the username/db_name are same
any error toasts showing up during/after schema population? - No error
how many schemas are expected, how many are showing? - except pgsql schema all other schemas are visible
is only a particular schema missing, or are others missing, too? -except pgsql schema all other schemas are visible

**Comment by villebro** [comment]
Highly interesting.. I will try to create a pgsql schema in my Postgres instance today to see if it's missing or not. I'll report back when I've tested.

**Comment by rusackas** [comment]
Is anyone still facing this? It's been silent for several months, so I'm tempted to close it as inactive.

**Comment by cnogradi** [comment]
Yes this issue has been plaguing us for months.  I just figured out that it seems to be related to the name pg* since our schema had a pgxxx prefix.  The schema appears in the catalog but whatever superset is doing it won't appear in the drop down selection in sql lab if it has this prefix.  In our case the work around is to rename.  This might be something particular to postgres but I am betting it's an issue in the superset code that removes pg* schemas from the list instead of removing pg_* schemas which are the ones reserved.

**Comment by rusackas** [comment]
Per the June 2025 comment from @cnogradi, this appears to affect any schema with a `pg` prefix, not just one named `pgsql`. It seems like Superset may be filtering out schemas starting with `pg` (which PostgreSQL uses internally for system schemas like `pg_catalog`, `pg_toast`). A fix would need to target the schema listing query to allow user-defined schemas with `pg` prefixes. PRs welcome if anyone can investigate.

**Comment by NavinAsokan** [comment]
I can take this PR and work on it. Please let me know the process to take this PR

**Comment by rusackas** [comment]
Found the root cause. It's not Superset's own filtering, it's SQLAlchemy's postgres dialect: `get_schema_names` excludes schemas matching `nspname NOT LIKE 'pg_%'`, and `_` in SQL LIKE is a single-char wildcard, so it silently drops anything starting with `pg` plus one more letter, not just the real `pg_` system schemas. An override of `get_schema_names` in `PostgresBaseEngineSpec` with a stricter check would fix it.

**Comment by rusackas** [comment]
I've got a PR open for this if anyone would be so kind as to test/review it :)
